Добавление нового функционала для хранения данных в БД

AlexanderKolnakov opened the pull request.

AlexanderKolnakov committed:
фиксация первых результатов

AlexanderKolnakov committed:
песты по фильмам пройдены

AlexanderKolnakov committed:
песты по фильмам и пользователям пройдены

AlexanderKolnakov committed:
последние правки

AlexanderKolnakov committed:
faet: - добавить новые слови контроллеров, сервисов и хранилища для Mpa и Genre.
- Добавить новую реализацию для бинов Storage у фильмов и пользователей, реализующую новый функционал по хранению данных в БД.
- создать БД в соответствии с ТЗ

EvgenyKoptubenko commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,74 +1,67 @@
     }
+
     @GetMapping("/films/{id}")
-    public Film findFilm(@PathVariable long id) throws IDException {
```
Если база данных пустая и сделать такой get запрос http://localhost:8080/films/22, то в ответ приходит 404 и stackTrace

`{
	"cause": null,
	"stackTrace": [
		{
			"classLoaderName": "app",
			"moduleName": null,
			"moduleVersion": null,
			"methodName": "error404",
			"fileName": "ErrorHandler.java",
			"lineNumber": 25,
			"nativeMethod": false,
			"className": "ru.yandex.practicum.filmorate.controller.error.ErrorHandler"
		},
		{
			"classLoaderName": null,
			"moduleName": "java.base",
			"moduleVersion": "17.0.4.1",
			"methodName": "invoke0",
			"fileName": "NativeMethodAccessorImpl.java",
			"lineNumber": -2,
			"nativeMethod": true,
			"className": "jdk.internal.reflect.NativeMethodAccessorImpl"
		},
		{`

EvgenyKoptubenko commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,74 +1,67 @@
 
-
     @PostMapping("/films")
     public Film create(@Valid @RequestBody Film film) throws ValidateException {
```
Если отправить такой запрос http://localhost:8080/films
{
	"name":"FilmName",
	"description":"descriptionItem"
}
То ожидается увидеть сообщение об ошибке, которое прописано в коде: Продолжительность фильма должна быть положительным целочисленным числом.

Но такое сообщение получается только в логе, а клиенту не приходит сообщение об ошибке.


Если к этому запросу добавить длительность
{
	"name":"FilmName",
	"description":"descriptionItem",
	"duration": 22
}
, то падает
java.lang.NullPointerException: Cannot invoke "ru.yandex.practicum.filmorate.model.Mpa.getId()" because "this.mpa" is null
	at ru.yandex.practicum.filmorate.model.Film.toMap(Film.java:63) ~[classes/:na]
	at ru.yandex.practicum.filmorate.storage.dao.FilmDbStorage.create(FilmDbStorage.java:66) ~[classes/:na]
	at ru.yandex.practicum.filmorate.service.FilmService.create(FilmService.java:41) ~[classes/:na]
	at ru.yandex.practicum.filmorate.controller.FilmController.create(FilmController.java:47) ~[classes/:na

EvgenyKoptubenko commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,74 +1,67 @@
         this.filmService = filmService;
     }
 
     @GetMapping("/films")
```
Слово films дублируется во всех методах контроллера - его лучше вынести и сделать общим для всех

EvgenyKoptubenko commented on src/main/java/ru/yandex/practicum/filmorate/storage/UserStorage.java:
```diff
@@ -8,8 +8,8 @@
-
     public User update(User user) throws ValidateException, IDException;
+    User getUser(int userID) throws IDException;
+    public boolean validateDataExists(int id);
```
в интерфейсе методы и так всегда public - можно убрать это слово

EvgenyKoptubenko commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/GenreDbStorage.java:
```diff
@@ -0,0 +1,49 @@
+
+    private final JdbcTemplate jdbcTemplate;
+    private final Logger log = LoggerFactory.getLogger(FilmDbStorage.class);
+    private static final String SQL_VALIDATE = "SELECT COUNT(*) AS count " +
```
count(*) лучше не использовать. Быстрее будет работать count(genre_id).

EvgenyKoptubenko reviewed (changes requested):
Работа сделана хорошо. Замечаний немного. поправь, пожалуйста.

AlexanderKolnakov committed:
refactor: - сократить повторяющийся путь в контроллерах при помощи аннотации @RequestMapping
fix: - исправить корректность возвращения сообщения об ошибке пользователю при некорректной передаче данных

AlexanderKolnakov committed:
fix: - актуализировать несколько тестов контроллеров

EvgenyKoptubenko reviewed (changes requested):
Если выкачать проект, запустить его, послать GET  http://localhost:8080/users/22
В ответе приходит стектрейс:
{
	"cause": null,
	"stackTrace": [
		{
			"classLoaderName": "app",
			"moduleName": null,
			"moduleVersion": null,
			"methodName": "error404",
			"fileName": "ErrorHandler.java",
			"lineNumber": 26,
			"nativeMethod": false,
			"className": "ru.yandex.practicum.filmorate.controller.error.ErrorHandler"
		},
		{
			"classLoaderName": null,
			"moduleName": "java.base",
			"moduleVersion": "17.0.4.1",
			"methodName": "invoke0",
			"fileName": "NativeMethodAccessorImpl.java",
			"lineNumber": -2,
			"nativeMethod": true,
			"className": "jdk.internal.reflect.NativeMethodAccessorImpl"
		},
		{
			"classLoaderName": null,
			"moduleName": "java.base",
			"moduleVersion": "17.0.4.1",
			"methodName": "invoke",
			"fileName": "NativeMethodAccessorImpl.java",
			"lineNumber": 77,
			"nativeMethod": false,
			"className": "jdk.internal.reflect.NativeMethodAccessorImpl"
		},
		{
			"classLoaderName": null,
			"moduleName": "java.base",
			"moduleVersion": "17.0.4.1",
			"methodName": "invoke",
			"fileName": "DelegatingMethodAccessorImpl.java",
			"lineNumber": 43,
			"nativeMethod": false,
			"className": "jdk.internal.reflect.DelegatingMethodAccessorImpl"
		},

AlexanderKolnakov committed:
fix: - исправить корректность ответа при ошибках

AlexanderKolnakov committed:
fix: - исправить mock тесты

EvgenyKoptubenko reviewed (approved):
Работа сделана хорошо. Спасибо.
